### PR TITLE
[search] Conscription numbers for Czech and Slovakia.

### PR DESCRIPTION
**Pull request by vng**
First patch to replace obsolete https://github.com/organicmaps/organicmaps/pull/1708

https://github.com/organicmaps/organicmaps/issues/579

Still a lot of work ahead.

### CODE REVIEW

**Comment by vng** [comment]
@andrewshadura @her01n
Can I remove "ev." prefixes from house numbers like here? https://www.openstreetmap.org/node/296611689
Or it has some sense and should be shown in OM _as-is_?

**Comment by andrewshadura** [comment]
In this case, ev. signifies it is a so-called provisional/Evidenz/evidenční number, not a conscription number. These are described in great detail in Slovak Wikipedia: https://sk.wikipedia.org/wiki/Eviden%C4%8Dn%C3%A9_%C4%8D%C3%ADslo_(budovy). These exist only in Czechia, and are used only on some building types (huts, small allotment houses, temporary buildings), and they rarely need addressing, so I think you can probably ignore them for now, and only use the street number.

Another important thing: always use the appropriate separate tag for the source of your house number, don't try to parse addr:housenumber. Do try the full text match on it though, so if someone types literally what's in there, they find the house regardless.

**Comment by vng** [comment]
Thanks for the clarification.

> Another important thing: always use the appropriate separate tag for the source of your house number, don't try to parse addr:housenumber. Do try the full text match on it though, so if someone types literally what's in there, they find the house regardless.

Well, separate HN tags at the current code state will produce a mess only. The idea of this PR is simply to treat '/' differently for Czech and Slovakia. House matching assumes to parse addr:housenumber anyway. Or I didn't get your idea well here?

**Comment by andrewshadura** [comment]
Well… It’s better than nothing, I guess, but a proper implementation needs to actually parse `addr:conscriptionnumber` and `addr:streetnumber`.

**Comment by vng** [comment]
Do you have some tricky OSM address examples in your mind that may not work?

**Comment by andrewshadura** [comment]
Any address where `addr:housenumber` is not in the format of `conscription/street` while `addr:conscriptionnumber` and `addr:streetnumber` are present.

**Comment by vng** [comment]
This address is searchable like "Trnavska 1355" (whole building) and "Trnavska 5" (particular entrance), but not like "Trnavska 1355/5" ("Trnavska 1355 5"). Not a big deal at first glance, but also fixable.

File: search/house_numbers_matcher.cpp [hunk 1]
```diff
@@ -577,6 +577,19 @@ bool HouseNumbersMatch(strings::UniString const & houseNumber, TokensT const & q
...
 }
 
+bool HouseNumbersMatchConscription(UniString const & houseNumber, TokensT const & queryParse)
+{
+  auto const beg = houseNumber.begin();
+  auto const end = houseNumber.end();
+  auto i = std::find(beg, end, '/');
+  if (i != end)
+  {
+    return HouseNumbersMatch(UniString(beg, i), queryParse) ||
```
Comment (biodranik) [inline search/house_numbers_matcher.cpp:587]: UniString_view? 😀

**Review by biodranik — APPROVED** [review]

File: search/features_layer_matcher.hpp [hunk 2]
```diff
@@ -116,8 +116,12 @@ class FeaturesLayerMatcher
     auto const interpol = ftypes::IsAddressInterpolChecker::Instance().GetInterpolType(feature);
     if (interpol != feature::InterpolType::None)
       return house_numbers::HouseNumbersMatchRange(feature.GetRef(), queryParse, interpol);
-    else
-      return house_numbers::HouseNumbersMatch(strings::MakeUniString(feature.GetHouseNumber()), queryParse);
+
+    auto const uniHouse = strings::MakeUniString(feature.GetHouseNumber());
+    if (feature.GetID().IsEqualCountry({"Czech", "Slovakia"}))
```
[the remark this answers is not in the document]
Comment (vng) [reply search/features_layer_matcher.hpp:121]: String literals are already constants or I didn't get here ..

File: search/features_layer_matcher.hpp [hunk 3]
```diff
@@ -116,8 +116,12 @@ class FeaturesLayerMatcher
...
+      return house_numbers::HouseNumbersMatchConscription(uniHouse, queryParse);
+
+    return house_numbers::HouseNumbersMatch(uniHouse, queryParse);
```
[the remark this answers is not in the document]
Comment (vng) [reply search/features_layer_matcher.hpp:124]: House number in general is a unicode string.

File: search/features_layer_matcher.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (biodranik) [reply search/features_layer_matcher.hpp:121]: StringIL is created each time, right?

[the remark this answers is not in the document]
Comment (vng) [reply search/features_layer_matcher.hpp:121]: initializer_list is something that is not used to say _created_, it has a fancy lifetime.
At least there is no need to bother here.

**Comment by vng** [comment]
Removed controversial commit with UniString-16 from this PR.

**Review by biodranik — APPROVED** [review]

**Pull request merged**